File: packages/go/canonicalfs/validate.go
package canonicalfs

import (
	"path"
	"strings"
)

func cleanRelative(rel string) (string, error) {
	if strings.ContainsRune(rel, '\x00') {
		return "", newError(ErrNULByte, "path contains NUL")
	}
	if rel == "" || rel == "." {
		return ".", nil
	}
	if isAbsolutePathLike(rel) {
		return "", newError(ErrAbsolutePath, "path must be relative to root")
	}
	if isDriveRelativePathLike(rel) {
		return "", newError(ErrDriveRelativePath, "Windows drive-relative paths are not accepted in CanonicalFS relative paths")
	}
	if strings.Contains(rel, "\\") {
		return "", newError(ErrOutsideRoot, "backslash separators are not accepted in CanonicalFS relative paths")
	}

	parts := make([]string, 0)
	for _, part := range strings.Split(rel, "/") {
		switch part {
		case "", ".":
			continue
		case "..":
			if len(parts) == 0 {
				return "", newError(ErrOutsideRoot, "path escapes root")
			}
			parts = parts[:len(parts)-1]
		default:
			parts = append(parts, part)
		}
	}
	if len(parts) == 0 {
		return ".", nil
	}
	return path.Join(parts...), nil
}

func isAbsolutePathLike(rel string) bool {
	if strings.HasPrefix(rel, "/") || strings.HasPrefix(rel, "\\\\") {
		return true
	}
	return len(rel) >= 3 && isASCIILetter(rel[0]) && rel[1] == ':' && (rel[2] == '/' || rel[2] == '\\')
}

func isDriveRelativePathLike(rel string) bool {
	return len(rel) >= 2 && isASCIILetter(rel[0]) && rel[1] == ':' && (len(rel) == 2 || rel[2] != '/' && rel[2] != '\\')
}

func isASCIILetter(b byte) bool {
	return b >= 'a' && b <= 'z' || b >= 'A' && b <= 'Z'
}

File: packages/ts/src/canonicalfs/validate.ts
import type { CanonicalRelativePath } from "../canonicalpath/types.js";
import { fsError } from "./errors.js";

export function validateRelativePath(rel: string): CanonicalRelativePath {
  if (rel.includes("\0")) throw fsError("ERR_NUL_BYTE", "path contains NUL");
  if (rel === "" || rel === ".") return "." as CanonicalRelativePath;
  if (isAbsolutePathLike(rel)) throw fsError("ERR_ABSOLUTE_PATH", "path must be relative to root");
  if (isDriveRelativePathLike(rel)) throw fsError("ERR_DRIVE_RELATIVE_PATH", "Windows drive-relative paths are not accepted in CanonicalFS relative paths");
  if (rel.includes("\\")) {
    throw fsError("ERR_OUTSIDE_ROOT", "backslash separators are not accepted in CanonicalFS relative paths");
  }

  const parts: string[] = [];
  for (const part of rel.split("/")) {
    if (part === "" || part === ".") continue;
    if (part === "..") {
      if (parts.length === 0) throw fsError("ERR_OUTSIDE_ROOT", "path escapes root");
      parts.pop();
      continue;
    }
    parts.push(part);
  }

  return (parts.length === 0 ? "." : parts.join("/")) as CanonicalRelativePath;
}

function isAbsolutePathLike(rel: string): boolean {
  if (rel.startsWith("/") || rel.startsWith("\\\\")) return true;
  return /^[A-Za-z]:[\\/]/.test(rel);
}

function isDriveRelativePathLike(rel: string): boolean {
  return /^[A-Za-z]:($|[^\\/])/.test(rel);
}
